packages/node-core/src/configure/ProjectUpgrade.service.ts

```diff
@@ -0,0 +1,296 @@
+    return this._projects.getAll();
+  }
+
+  get currentProject(): P {
+    return this.#currentProject;
+  }
+
+  get currentHeight(): number {
+    return this.#currentHeight;
+  }
+
+  async setCurrentHeight(height: number): Promise<void> {
+    this.#currentHeight = height;
+    const newProjectDetails = this._projects.getDetails(height);
+    assert(newProjectDetails, `Unable to find project for height ${height}`);
+    const {startHeight, value: newProject} = newProjectDetails;
+
+    const hasChanged = this.#currentProject !== newProject;
+    // Need to set this so that operations under hasChanged use the new project
+    this.#currentProject = newProject;
+
+    if (hasChanged) {
+      try {
+        // Use the project start height here for when resuming indexing at an arbitrary height
+        await this.updateIndexedDeployments(newProject.id, startHeight);
+      } catch (e: any) {
+        logger.error(e, 'Failed to update deployment metadata');
+        process.exit(1);
+      }
+
+      try {
+        await this.onProjectUpgrade?.(startHeight, newProject);
+      } catch (e: any) {
+        logger.error(e, `Failed to complete upgrading project`);
+        process.exit(1);
+      }
+
+      logger.info(`Project upgraded to ${newProject.id} at height ${height}`);
+    }
+  }
+
+  static async create<P extends ISubqueryProject>(
+    startProject: P, // The project passed in via application start
+    loadProject: (ipfsCid: string) => Promise<P>,
+    startHeight?: number // How far back we need to load parent versions
+  ): Promise<ProjectUpgradeSevice<P>> {
+    const projects: Map<number, P> = new Map();
+
+    let currentProject = startProject;
+    let currentHeight: number | undefined;
+
+    const addProject = (height: number, project: P) => {
+      this.validateProject(startProject, project);
+      projects.set(height, project);
+    };
+
+    // eslint-disable-next-line no-constant-condition
+    while (true) {
+      const projectStartHeight = getStartHeight(currentProject.dataSources);
+
+      // Set the current height to the start of the startProject if not provided
+      if (currentHeight === undefined) {
+        currentHeight = startProject.parent?.block ?? projectStartHeight;
+      }
+
+      // At the end of the chain
+      if (!currentProject.parent) {
+        // At the limit of the start height for DS
+        if (startHeight !== undefined && startHeight < projectStartHeight) {
+          break;
+        }
+        addProject(projectStartHeight, currentProject);
+        break;
+      }
+      if (currentProject.parent.block) {
+        addProject(currentProject.parent.block, currentProject);
+      }
+      // At the limit of the start height for Parent
+      if (startHeight !== undefined && startHeight >= currentProject.parent.block) {
+        break;
+      }
+
+      // Load the next project and repeat
+      const nextProject = await loadProject(currentProject.parent.reference);
```

Comment:
Is that possible when we load the parent reference failed?

Reply:
I have a better error message now